api/src/main/java/marquez/db/RunStateDao.java

```diff
@@ -15,71 +15,40 @@
 import marquez.db.models.RunStateRow;
-import org.jdbi.v3.sqlobject.CreateSqlObject;
-import org.jdbi.v3.sqlobject.SqlObject;
 import org.jdbi.v3.sqlobject.config.RegisterRowMapper;
 import org.jdbi.v3.sqlobject.statement.SqlQuery;
 import org.jdbi.v3.sqlobject.transaction.Transaction;
 
 @RegisterRowMapper(RunStateRowMapper.class)
-public interface RunStateDao extends SqlObject {
-  @CreateSqlObject
-  DatasetDao createDatasetDao();
-
-  @CreateSqlObject
-  RunDao createRunDao();
-
-  @Transaction
-  default void insert(
-      RunStateRow row,
-      List<UUID> outputVersionUuids,
-      boolean starting,
-      boolean done,
-      boolean complete) {
-    withHandle(
-        handle ->
-            handle
-                .createUpdate(
-                    "INSERT INTO run_states (uuid, transitioned_at, run_uuid, state)"
-                        + "VALUES (:uuid, :transitionedAt, :runUuid, :state)")
-                .bindBean(row)
-                .execute());
-    // State transition
-    final Instant updateAt = row.getTransitionedAt();
-    createRunDao().updateRunState(row.getRunUuid(), updateAt, row.getState());
-    if (starting) {
-      createRunDao().updateStartState(row.getRunUuid(), updateAt, row.getUuid());
-    }
-    if (done) {
-      createRunDao().updateEndState(row.getRunUuid(), updateAt, row.getUuid());
-    }
-    // Modified
-    if (complete && outputVersionUuids != null && outputVersionUuids.size() > 0) {
-      createDatasetDao().updateLastModifedAt(outputVersionUuids, updateAt);
-    }
-  }
-
+public interface RunStateDao extends MarquezDao {
   @SqlQuery("SELECT * FROM run_states WHERE uuid = :rowUuid")
   Optional<RunStateRow> findBy(UUID rowUuid);
 
-  @SqlQuery(
-      "SELECT * FROM run_states "
-          + "WHERE run_uuid = :runUuid "
-          + "ORDER BY transitioned_at DESC "
-          + "LIMIT 1")
-  Optional<RunStateRow> findLatest(UUID runUuid);
-
   @SqlQuery("SELECT COUNT(*) FROM run_states")
   int count();
 
   @SqlQuery(
       "INSERT INTO run_states (uuid, transitioned_at, run_uuid, state)"
           + "VALUES (:uuid, :now, :runUuid, :runStateType) RETURNING *")
   RunStateRow upsert(UUID uuid, Instant now, UUID runUuid, RunState runStateType);
+
+  @Transaction
+  default void updateRunState(UUID runUuid, RunState runState, Instant transitionedAt) {
+    RunDao runDao = createRunDao();
+    RunStateRow runStateRow = upsert(UUID.randomUUID(), transitionedAt, runUuid, runState);
+    runDao.updateRunState(runUuid, transitionedAt, runState);
+    if (runState.isDone()) {
+      runDao.updateEndState(runUuid, transitionedAt, runStateRow.getUuid());
+    } else if (runState.isStarting()) {
+      runDao.updateStartState(runUuid, transitionedAt, runStateRow.getUuid());
+    }
+    //    if (complete && outputVersionUuids != null && outputVersionUuids.size() > 0) {
+    //      createDatasetDao().updateLastModifedAt(outputVersionUuids, updateAt);
+    //    }
```

Comment:
what is the logic supposed to be for last modified at?

Reply:
I would think that it matches the end time of the last run the had this dataset as an output?

Reply:
@wslulciuc ?

Reply:
@henneberger: When a job reaches a `COMPLETE` state, there's a reasonable assumption that the output datasets have been modified. We can set `modifedAt` equal to `transitionedAt`